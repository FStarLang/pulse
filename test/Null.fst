module Null

#lang-pulse
open Pulse
open Pulse.Lib.Box
module Box = Pulse.Lib.Box
module Ref = Pulse.Lib.Reference

(*
PRESERVES(null_or_live(x));
int foo(int *x) {
  if (x == NULL) {
    return 0;
  } else {
    return *x;
  }
}
*)

let x : ref int = Pulse.Lib.Reference.null
let y () : box int = Box.null

fn foo ()
  returns r : box int
{
  Box.null
}

// This should maybe go into the Pulse library or a prelude.
unfold
let live #a (r : ref a) =
  exists* v. r |-> v

// Idem
unfold
let null_or_live #a (r : ref a) =
  if Ref.is_null r
  then emp
  else live r

fn test (x : ref int)
  preserves null_or_live x
  returns int
{
  if (Ref.is_null x) {
    rewrite emp as null_or_live x;
    0
  } else {
    let res = Ref.(!x);
    rewrite live x as null_or_live x;
    res
  }
}
