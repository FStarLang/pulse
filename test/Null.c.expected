/* krml header omitted for test repeatability */


#include "Null.h"

krml_checked_int_t *Null_x = NULL;

krml_checked_int_t *Null_y(void)
{
  return NULL;
}

krml_checked_int_t *Null_foo(void)
{
  return NULL;
}

krml_checked_int_t Null_test(krml_checked_int_t *x1)
{
  if (x1 == NULL)
    return (krml_checked_int_t)0;
  else
    return *x1;
}

